File: src/plugins/math_matvec.cpp
// Implement the matrix/vector authoring helpers over the core math grid.
#include "plugins/math_matvec.hpp"

#include <limits>
#include <stdexcept>
#include <utility>
#include <vector>

namespace dans::document::plugins
{
auto Math::MatVec::matrix_from_rows(std::vector<Row> rows) -> Math
{
    if (rows.empty())
    {
        throw std::invalid_argument{"A matrix requires at least one row"};
    }
    const auto column_count = rows.front().cells_.size();
    if (column_count == usize{0})
    {
        throw std::invalid_argument{"A matrix row cannot be empty"};
    }
    if (column_count > std::numeric_limits<usize>::max() / rows.size())
    {
        throw std::invalid_argument{"Matrix dimensions exceed the supported size"};
    }

    std::vector<Math> cells{};
    cells.reserve(rows.size() * column_count);
    for (auto& row : rows)
    {
        if (row.cells_.size() != column_count)
        {
            throw std::invalid_argument{"Every matrix row must have the same number of cells"};
        }
        for (auto& cell : row.cells_)
        {
            cells.push_back(std::move(cell));
        }
    }
    return bracket(Math::grid(rows.size(), column_count, std::move(cells)));
}

auto Math::MatVec::bracket(Math grid) -> Math
{
    auto result = Math::delimited(Math::Delimiter::square);
    result.body(std::move(grid));
    return result;
}
}  // namespace dans::document::plugins

File: builder/src/plugins/mathMatVec.ts
// Optional matrix/vector authoring extension built entirely from core math primitives.
import type { MathEditorExtension } from "../math/editorExtension";
import {
  createMathDelimited,
  createMathGrid,
  type MathExpression,
} from "../model/math";

function bracketGrid(
  rows: number,
  columns: number,
  cells?: readonly MathExpression[],
): MathExpression {
  return createMathDelimited("brackets", createMathGrid(rows, columns, cells));
}

export function createMathMatrix(
  rows: readonly (readonly MathExpression[])[],
): MathExpression {
  const firstRow = rows[0];
  if (firstRow === undefined || firstRow.length === 0) {
    throw new Error("A matrix requires at least one non-empty row");
  }
  const columns = firstRow.length;
  if (!rows.every((row) => row.length === columns)) {
    throw new Error("Every matrix row must have the same number of cells");
  }
  return bracketGrid(rows.length, columns, rows.flat());
}

export function createEmptyMathMatrix(rows: number, columns: number): MathExpression {
  return bracketGrid(rows, columns);
}

export function createMathRowVector(
  cells: readonly MathExpression[],
): MathExpression {
  if (cells.length === 0) {
    throw new Error("A row vector requires at least one cell");
  }
  return bracketGrid(1, cells.length, cells);
}

export function createMathColumnVector(
  cells: readonly MathExpression[],
): MathExpression {
  if (cells.length === 0) {
    throw new Error("A column vector requires at least one cell");
  }
  return bracketGrid(cells.length, 1, cells);
}

export const mathMatVecEditorExtension: MathEditorExtension = Object.freeze({
  id: "dans.math.matvec",
  label: "Matrices and vectors",
  items: Object.freeze([
    Object.freeze({
      id: "matvec-matrix-2x2",
      label: "2×2",
      description: "Square two-by-two matrix",
      create: () => createEmptyMathMatrix(2, 2),
    }),
    Object.freeze({
      id: "matvec-matrix-2x3",
      label: "2×3",
      description: "Rectangular two-by-three matrix",
      create: () => createEmptyMathMatrix(2, 3),
    }),
    Object.freeze({
      id: "matvec-row-vector",
      label: "[···]",
      description: "Three-component row vector",
      create: () => createEmptyMathMatrix(1, 3),
    }),
    Object.freeze({
      id: "matvec-column-vector",
      label: "[⋮]",
      description: "Three-component column vector",
      create: () => createEmptyMathMatrix(3, 1),
    }),
  ]),
});
